lib/utils/ignore.ts

```diff
@@ -294,45 +332,57 @@ export async function filterFilesWithDockerignore(
 	const dockerignoreServiceDirs: string[] = multiDockerignore
 		? Object.keys(ignoreByDir).filter((dir) => dir && dir !== root)
 		: [];
+	const files = await listFiles(projectDir);
+
 	const dockerignoreFiles: FileStats[] = [];
 	const filteredFileList: FileStats[] = [];
-	const klaw = await import('klaw');
-
-	await new Promise((resolve, reject) => {
-		// Looking at klaw's source code, `preserveSymlinks` appears to only
-		// afect the `stats` argument to the `data` event handler
-		klaw(projectDir, { preserveSymlinks: false })
-			.on('error', reject)
-			.on('end', resolve)
-			.on('data', (item: { path: string; stats: Stats }) => {
-				const { path: filePath, stats } = item;
-				// With `preserveSymlinks: false`, filePath cannot be a symlink.
-				// filePath may be a directory or a regular or special file
-				if (!stats.isFile()) {
+	await Promise.all(
+		files.map(async (filePath) => {
```

Comment:
If I am reading it right, this will create an array of promises, the array size being equal to the number of files in the project, and then all promises are awaited in parallel. If the project had 500,000 files, this would produce 500,000 promises awaiting in parallel.

I am concerned about memory usage. Consider also some people's hack week projects where they wish to run `balena push` in low-memory IoT devices for the purpose of running an IDE on the device itself.

The following StackOverflow post suggests a hard Node.js limit of 2.1 million promises awaited in parallel, in which case a project with just over 2.1 million files would necessarily fail, even if RAM was not a problem: https://stackoverflow.com/questions/55753746/how-much-is-the-limit-of-promise-all

There is probably a number of concurrent promises above which further parallelisation yields no further performance improvement. I mean, running perhaps "100" (?) `fs.stat` calls in parallel might be faster than running them sequentially, but running "500,000" in parallel might be no faster than running "100" at a time. In fact, I would not be surprised if, as the number got large enough, performance worsened due to extra "bookkeeping" (epoll array iterations looking for fd's that are ready?) or in case the system had to resort to memory pagination in swap space (disk).

Bluebird has a fancy `promise.map` implementation that accepts a `concurrency` parameter:
> http://bluebirdjs.com/docs/api/promise.map.html
> The concurrency limit applies to Promises returned by the mapper function and it basically limits the number of Promises created. For example, if concurrency is 3 and the mapper callback has been called enough so that there are three returned Promises currently pending, no further callbacks are called until one of the pending Promises resolves. So the mapper function will be called three times and it will be called again only after at least one of the Promises resolves.

Perhaps you could test some concurrency numbers and find the sweet one that will hopefully be relatively low.

Reply:
@pdcastro This function will walk one layer of directories at a time, so even on the Jellyfish project, which is a very large multicontainer app, the maximum number of promises created is 8993.
Switching to `Bluebird.map` is a convenient way to prevent something crazy happening, and setting the concurrency to 750 has no impact on performance.

Reply:
> This function will walk one layer of directories at a time

Which function, `listFiles`? But it is being called as follows:
```ts
    const files = await listFiles(projectDir);
```
So even if `listFiles` builds the array on one layer of directories at a time, the `files` array above will be as large as all files in the project, no?

> setting the concurrency to 750 has no impact on performance.

So, shouldn't we at least do that?

Reply:
@pdcastro I should have been clearer, I've set the concurrency to 750 for listFiles, We can do something similar for the `ignore` processing as well.

Reply:
@pdcastro I did some more testing with both `listFiles` and the processing of the returned values and found that using `Bluebird` instead of native promises was significantly slower.
Using `Bluebird.map` without concurrency completed the tar process in 12387ms, after switching to native promises + [_.chunk](https://lodash.com/docs/4.17.15#chunk) to limit concurrency the process completed in 9057ms.
The actual tarballing step takes ~2s on my test setup, so it seems that using Bluebird is ~47% slower than a "native" approach!
I've now refactored both places I was using `Promise.all` to first chunk the arrays at a sensible size to prevent arbitrarily large file systems causing high memory consumption.